### PR TITLE
[build] Add CodeQL Security Scan workflow

### PR DESCRIPTION
**Pull request by huangminchn**
## Summary, imperative, start upper case, don't end with a period
Add a workflow to run CodeQL security scan every Monday,Wednesday,Friday evening.

## How was this PR tested?
Tested in my fork: https://github.com/huangminchn/venice/actions/workflows/codeql.yml

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: .github/workflows/codeql.yml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+# ******** NOTE ********
+# We have attempted to detect the languages in your repository. Please check
+# the `language` matrix defined below to confirm you have the correct set of
+# supported CodeQL languages.
+#
+name: "CodeQL"
+
+on:
+  push:
+    branches: [ "main", hotfix-[0-9]*.[0-9]*.[0-9]*, release-* ]
```
Comment (nisargthakkar) [inline .github/workflows/codeql.yml:16]: Do we need to define the branch patterns? I'm not happy with our current wildcard patterns and plan to rework them at some point

Can we run it on all branches? Skip the arg or use `*`?

File: .github/workflows/codeql.yml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  pull_request:
+    # The branches below must be a subset of the branches above
+    branches: [ "main" ]
+  # At 11:27 PM, only on Monday, Wednesday, and Friday
+  schedule:
+    - cron: '27 23 * * 1,3,5'
```
Comment (adamxchen) [inline .github/workflows/codeql.yml:22]: how are we going to use the scan report? Can't we use it for every PR like [this](https://docs.github.com/en/code-security/code-scanning/automatically-scanning-your-code-for-vulnerabilities-and-errors/configuring-code-scanning-for-a-repository#understanding-the-pull-request-checks)?

**Review by adamxchen — COMMENTED** [review]
Thank you, Min for being bias on actions!

**Comment by sixpluszero** [comment]
What does this CodelQL scan do? From Google search looks like it is similar spot bug?

File: .github/workflows/codeql.yml (hunk 2 above)
Comment (nisargthakkar) [reply .github/workflows/codeql.yml:22]: Is there any significance of the `27`? Can we make it `0`? 
(Just for OCD 😅)

Comment (nisargthakkar) [reply .github/workflows/codeql.yml:22]: Regarding @adamxchen's comment, we can make it a required check for merging a PR once this change is merged in

File: .github/workflows/codeql.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
+# For most projects, this workflow file will not need changing; you simply need
```
Comment (gaojieliu) [inline .github/workflows/codeql.yml:1]: Can we include a link to the original copy?
I assume we copied this file from some other places.

File: .github/workflows/codeql.yml (hunk 2 above)
Comment (nisargthakkar) [reply .github/workflows/codeql.yml:22]: This one already runs on all PRs along with the schedule. The reason for the schedule is that vulnerabilities might get marked as such after a PR that pulled them in is merged.

File: .github/workflows/codeql.yml (hunk 3 above)
Comment (nisargthakkar) [reply .github/workflows/codeql.yml:1]: It is auto generated by GitHub. (Settings -> Code security and analysis -> CodeQL analysis -> Set up -> Advanced)

Comment (gaojieliu) [reply .github/workflows/codeql.yml:1]: I see.
Then we are good.

**Review by nisargthakkar — DISMISSED** [review]

**Comment by adamxchen** [comment]
@huangminchn  Hey Min, this PR hasn't been inactive for a while so let me close it so our review queue could be cleaner. Please re-open it when you have the bandwidth working on it again.

**Pull request closed without merging**